src/main/java/org/mastodon/mamut/tomancak/lineage_registration/LineageTreeUtils.java

```diff
@@ -0,0 +1,51 @@
+package org.mastodon.mamut.tomancak.lineage_registration;
+
+import org.mastodon.collection.RefSet;
+import org.mastodon.collection.ref.RefSetImp;
+import org.mastodon.mamut.model.ModelGraph;
+import org.mastodon.mamut.model.Spot;
+import org.mastodon.pool.PoolCollectionWrapper;
+
+public class LineageTreeUtils
+{
+	/**
+	 * @return the set of root nodes of the give graph.
+	 */
+	public static RefSet<Spot> getRoots( ModelGraph graph ) {
+		PoolCollectionWrapper<Spot> vertices = graph.vertices();
+		RefSetImp<Spot> roots = new RefSetImp<>( vertices.getRefPool() );
+		for( Spot spot : vertices )
+			if(spot.incomingEdges().isEmpty())
+				roots.add(spot);
+		return roots;
+	}
+
+	/**
+	 * The given spot belongs to a branch. This method returns the last
+	 * spot of this branch.
+	 */
+	public static Spot getBranchEnd( ModelGraph graph, final Spot spot )
+	{
+		Spot s = graph.vertexRef();
+		s.refTo( spot );
+		while(s.outgoingEdges().size() == 1) {
+			s = s.outgoingEdges().get( 0 ).getTarget(s);
+		}
+		return s;
+	}
+
+	/**
+	 * @return true if the given spot is part of a branch that divides.
+	 */
+	public static boolean doesBranchDivide( ModelGraph graph, final Spot spot )
+	{
+		Spot branchEnd = getBranchEnd( graph, spot );
+		try
+		{
+			return branchEnd.outgoingEdges().size() > 1;
+		}
+		finally {
+			graph.releaseRef( branchEnd );		
+		}
+	}
+}
```

Comment:
this was a general comment for the whole class

Reply:
I remember @tinevez once telling me, that it should now be impossible to create an edge in reverse order in Mastodon.

Should we maybe write a plugin that checks a graph for these kind of mistakes and allows to automatically fix them?

Reply:
Yes, please. I would say we need to have such a plugin and would even say it should be executed always as part of the data-loading process. That way, all plugins can rely on the incoming/outgoing semantics, even for old data.

In my earlier days I am very sure the Mastodon was working as I said before, I tested that, and I had to be using [complex tests](https://github.com/mastodon-sc/mastodon-tomancak/blob/c6d13d1f4e87096141157121f387786f74a9a8e4/src/main/java/org/mastodon/mamut/tomancak/util/SpotsIterator.java#L274-L292); but I recall it was mentioned that this may have got fixed, I don't know for sure... in any case, there might still be floating around some old data

The fixed/correct version is clearly a lot better option to have (easier -> faster tests is super important when data grow large)

@tinevez what do you think if such a sanity checker would be part of the loading? should it warn and ask users to run fixer plugin, or fix on the fly?

Reply:
Implementing the plugin is not in the scope of this PR. But I created an issue so we don't forget about the idea: https://github.com/mastodon-sc/mastodon-tomancak/issues/12

Reply:
Good that you created this issue.